### PR TITLE
feat: Take IdentifierProvider into use for equality determination in list-box

### PR DESCRIPTION
**Pull request by taefi**
Take IdentifierProvider into use for equality determination in `ListBox` and `MultiSelectListBox` by overriding `valueEquals`:
Fix: #225 
Fix: #227

### CODE REVIEW

File: vaadin-list-box-flow-parent/vaadin-list-box-flow/src/test/java/com/vaadin/flow/component/listbox/test/ListBoxUnitTest.java [hunk 1]
```diff
@@ -48,6 +53,40 @@ public void setItemEnabledProvider_itemDisabled() {
...
+        // Make the names similar to the name of not selected one to mess
+        // with the <equals> implementation in CustomItem:
+        first.setName("Second");
+        listDataView.refreshItem(first);
+        third.setName("Second");
+        listDataView.refreshItem(third);
+
+        // Select the item not with the reference of existing item, but instead
+        // with just the Id:
+        listBox.setValue(new CustomItem(2L));
```
Comment (mshabarov) [inline vaadin-list-box-flow-parent/vaadin-list-box-flow/src/test/java/com/vaadin/flow/component/listbox/test/ListBoxUnitTest.java:84]: What if we set `new CustomItem(2L, "Second")` to the list box? Or even `new CustomItem("Second")`? Should we test this case also?

File: vaadin-list-box-flow-parent/vaadin-list-box-flow/src/main/java/com/vaadin/flow/component/listbox/MultiSelectListBox.java [hunk 2]
```diff
@@ -133,4 +133,18 @@ public Registration addSelectionListener(
                 .selectionChange(new MultiSelectionEvent<>(this, this,
                         event.getOldValue(), event.isFromClient())));
     }
+
+    @Override
+    protected boolean valueEquals(Set<T> value1, Set<T> value2) {
```
Comment (mshabarov) [inline vaadin-list-box-flow-parent/vaadin-list-box-flow/src/main/java/com/vaadin/flow/component/listbox/MultiSelectListBox.java:138]: Same comment about javadoc here

File: vaadin-list-box-flow-parent/vaadin-list-box-flow-integration-tests/src/test/java/com/vaadin/flow/component/listbox/test/ListBoxDataViewIT.java [hunk 3]
```diff
@@ -223,28 +227,58 @@ public void testListDataView_nextAndPreviousItem_shouldComplyWithOriginalOrder()
...
         Assert.assertEquals("second rendered radio button's text should be",
-                "second", buttons.get(1).getText());
+                "second", items.get(1).getText());
         Assert.assertEquals("third rendered radio button's text should be",
-                "third", buttons.get(2).getText());
+                "third", items.get(2).getText());
+    }
+
+    @Test
+    public void setIdentifierProvider_forMultiSelectListBox_shouldSelectCorrectItems_dueToIdentifier() {
```
Comment (mshabarov) [inline vaadin-list-box-flow-parent/vaadin-list-box-flow-integration-tests/src/test/java/com/vaadin/flow/component/listbox/test/ListBoxDataViewIT.java:256]: Rework the naming https://sites.google.com/a/vaadin.com/fw-flow-team-s-wiki/contributing-code/flow-way-of-devloping

File: vaadin-list-box-flow-parent/vaadin-list-box-flow-integration-tests/src/test/java/com/vaadin/flow/component/listbox/test/ListBoxDataViewIT.java [hunk 4]
```diff
@@ -223,28 +227,58 @@ public void testListDataView_nextAndPreviousItem_shouldComplyWithOriginalOrder()
...
+        findElement(By.id(MULTI_SELECT_LIST_BOX_SELECTION_UPDATE_BUTTON)).click();
+
+        selectedIdsSpan = findElement(
+                By.id(MULTI_SELECT_LIST_BOX_SELECTED_IDS_SPAN));
+        Assert.assertEquals("Selected item ids should be", "2, 4",
+                selectedIdsSpan.getText());
+    }
+
+    @Test
+    public void setIdentifierProvider_forListBox_shouldSelectCorrectItem_dueToIdentifier() {
```
Comment (mshabarov) [inline vaadin-list-box-flow-parent/vaadin-list-box-flow-integration-tests/src/test/java/com/vaadin/flow/component/listbox/test/ListBoxDataViewIT.java:271]: Rework the naming https://sites.google.com/a/vaadin.com/fw-flow-team-s-wiki/contributing-code/flow-way-of-devloping

**Review by mshabarov — CHANGES_REQUESTED** [review]

File: vaadin-list-box-flow-parent/vaadin-list-box-flow/src/main/java/com/vaadin/flow/component/listbox/ListBox.java [hunk 5]
```diff
@@ -64,4 +64,13 @@ public ListBox() {
                 .findFirst().orElseThrow(() -> new IllegalArgumentException(
                         "Could not find given value from the item set"));
     }
+
+    @Override
+    protected boolean valueEquals(T value1, T value2) {
```
[the remark this answers is not in the document]
Comment (taefi) [reply vaadin-list-box-flow-parent/vaadin-list-box-flow/src/main/java/com/vaadin/flow/component/listbox/ListBox.java:85]: Agree! Done.

**Review by mshabarov — CHANGES_REQUESTED** [review]

**Comment by mshabarov** [comment]
A couple of typos, but LGTM

**Review by mshabarov — APPROVED** [review]

**Pull request merged**